Public/Src/Cache/MemoizationStore/Distributed/Stores/RedisMemoizationDatabase.cs

```diff
@@ -0,0 +1,192 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Threading;
+using BuildXL.Cache.ContentStore.Distributed.Redis;
+using BuildXL.Cache.ContentStore.Distributed.NuCache;
+using BuildXL.Cache.ContentStore.Interfaces.Extensions;
+using BuildXL.Cache.ContentStore.Interfaces.Results;
+using BuildXL.Cache.ContentStore.Interfaces.Time;
+using BuildXL.Cache.ContentStore.Tracing;
+using BuildXL.Cache.ContentStore.Tracing.Internal;
+using BuildXL.Cache.MemoizationStore.Interfaces.Results;
+using BuildXL.Cache.MemoizationStore.Interfaces.Sessions;
+using BuildXL.Cache.MemoizationStore.Stores;
+using BuildXL.Utilities;
+using StackExchange.Redis;
+using BuildXL.Cache.ContentStore.UtilitiesCore.Internal;
+using System.Security.Cryptography;
+
+namespace BuildXL.Cache.MemoizationStore.Distributed.Stores
+{
+    /// <summary>
+    /// Implements a memoization database through Redis.
+    /// </summary>
+    internal class RedisMemoizationDatabase : MemoizationDatabase

```

Comment:
Question: there seems not to be any garbage collection mechanism for Redis. Is it not needed?

Reply:
Everything in Redis has TTL and will be automatically garbage collected

Reply:
AFAIK, that is not true. We need to set the TTL in the batch which performs the CompareExchange.

---
In reply to: [320995030](https://github.com/microsoft/BuildXL/pull/832#discussion_r320995030) [](ancestors = 320995030)

Reply:
Assigning a TTL to keys via KeyExpire handle garbage collection.

---
In reply to: [320977839](https://github.com/microsoft/BuildXL/pull/832#discussion_r320977839) [](ancestors = 320977839)